File: plugins/top10-scan/test-fixtures/vulnerable-aspnet/Controllers/AdminController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace VulnerableApp.Controllers;

[ApiController]
[Route("api/[controller]")]
// A01: No [Authorize] attribute - anyone can access admin endpoints
public class AdminController : ControllerBase
{
    private readonly AppDbContext _db;

    public AdminController(AppDbContext db)
    {
        _db = db;
    }

    [HttpGet("users")]
    public async Task<IActionResult> ListAllUsers()
    {
        // Returns all user data including password hashes
        var users = await _db.Users.ToListAsync();
        return Ok(users);
    }

    [HttpDelete("users/{id}")]
    public async Task<IActionResult> DeleteUser(int id)
    {
        var user = await _db.Users.FindAsync(id);
        if (user == null)
            return NotFound();

        _db.Users.Remove(user);
        await _db.SaveChangesAsync();

        // A09-adjacent: No audit logging of admin deletion
        return Ok(new { message = "Deleted" });
    }
}

File: plugins/top10-scan/test-fixtures/vulnerable-php/app/Http/Controllers/AdminController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class AdminController extends Controller
{
    // A01: No auth check, no admin role verification
    public function listUsers()
    {
        // Returns all users including password hashes
        $users = DB::table('users')->get();
        return response()->json($users);
    }

    // A01: No auth check on destructive action
    // A09: No audit logging of user deletion
    public function deleteUser($id)
    {
        DB::table('users')->where('id', $id)->delete();
        return response()->json(['message' => 'User deleted']);
    }

    // A01: No auth check on sensitive logs
    public function viewLogs()
    {
        $logs = DB::table('audit_logs')->get();
        return response()->json($logs);
    }
}
